### PR TITLE
Veeam-vm-backup function added

### PR DESCRIPTION
**Pull request by mirela-manoleva**
## Summary

Please describe your changes and what you plan to introduce as part of this pull request submission. If it fixes a bug or resolves a feature request, be sure to link to the issue numbers following the template below.

## Pull Request Checklist
🚨 Please review the [guidelines for contributing](https://vmweventbroker.io/community) to this repository.

- [x] Please ensure that you are making a pull request against the **Development** branch
- [x] Please use the `WIP` keyword in the title of your PR if you are not ready for review
- [x] Please ensure that you have opened a Github Issue if you are resolving/fixing a problem
- [x] Please ensure that you have [signed](https://help.github.com/en/github/authenticating-to-github/signing-commits) all commits and that you have [squashed](https://medium.com/@slamflipstrom/a-beginners-guide-to-squashing-commits-with-git-rebase-8185cf6e62ec) all relevant commmits related to your change
- [x] Please make sure that you have tested your change locally by successfully [building and deploying the VMware Event Broker Appliance](https://vmweventbroker.io/kb/contribute-appliance) and/or [building and deploying VMware Event Router](https://vmweventbroker.io/kb/contribute-eventrouter)
- [x] Please include any relevant screenshots and/or output as part of your testing
- [x] Please include any documentation updates that is applicable for your changes

## Change Type

What types of changes does your code introduce to the VMware Event Broker Appliance?

_Put an `x` in all boxes that apply_

Please check the type of change your PR introduces:
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] Feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation changes
- [ ] Other (please describe):

## Resolved Issues

List of Issues closed or resolved by this PR. Add multiple `Closes` keyword followed by the issue number (e.g. Closes #ISSUE-NUMBER)

* Closes #ISSUE-NUMBER

## Testing Verification

* Short summary of testing (e.g. successfully built and deployed VMware Event Broker Appliance)
* Please include any relevant screenshots and/or output as part of the testing

## Additional Information

* Any other details you wish to include or mention

If you have any questions/comments, feel free to reach out to team on Slack [#vcenter-event-broker-appliance](https://vmwarecode.slack.com/archives/CQLT9B5AA)

Thank you from the VEBA Team! 🥳

### CODE REVIEW

File: examples/java/veeam-vm-backup/README.MD
Comment (embano1) [inline examples/java/veeam-vm-backup/README.MD:84]: Is this a special case for the Java template where you have to copy around content from `template`?

File: examples/java/veeam-vm-backup/README.MD
Comment (embano1) [inline examples/java/veeam-vm-backup/README.MD:5]: `mutation`?

**Review by embano1 — COMMENTED** [review]
Thank you Mirela! Can you please complete the PR form above, e.g. check each checkbox that applies (incl. the PR checklist)? Also, can you provide details on how you tested this, e.g. which versions of VEBA, or router, vCenter, OpenFaaS, etc. in case you use standalone deployments?

File: examples/java/veeam-vm-backup/README.MD
Comment (embano1) [inline examples/java/veeam-vm-backup/README.MD:67]: Would the example also work with the `java11` template which is not marked deprecated and uses openjdk? https://github.com/openfaas/templates/tree/master/template/java11

File: examples/java/veeam-vm-backup/veeam-vm-backup.yml
Comment (embano1) [inline examples/java/veeam-vm-backup/veeam-vm-backup.yml:9]: Can you please change this to `vmware/veba-java-veeam-vm-backup:latest`

**Comment by mirela-manoleva** [comment]
> Thank you Mirela! Can you please complete the PR form above, e.g. check each checkbox that applies (incl. the PR checklist)? Also, can you provide details on how you tested this, e.g. which versions of VEBA, or router, vCenter, OpenFaaS, etc. in case you use standalone deployments?

The forms are completed now. As for the versions - the function is tested on VEBA 0.3.0 and vCenter 7.0u1, also OpenFaaS 0.12.13 locally.

File: examples/java/veeam-vm-backup/README.MD
Comment (mirela-manoleva) [reply examples/java/veeam-vm-backup/README.MD]: This part is removed from the edited README.md file, as these steps were not needed for the successful deployment of the function.

File: examples/java/veeam-vm-backup/README.MD
Comment (mirela-manoleva) [reply examples/java/veeam-vm-backup/README.MD]: This refers to state changed event - it is edited in the updated documentation.

File: examples/java/veeam-vm-backup/README.MD
Comment (mirela-manoleva) [reply examples/java/veeam-vm-backup/README.MD]: The function works with the java11 OpenFaaS template - details on this are added in the updated documentation.

File: examples/java/veeam-vm-backup/veeam-vm-backup.yml
Comment (mirela-manoleva) [reply examples/java/veeam-vm-backup/veeam-vm-backup.yml]: This is done in the updated files.

**Review by vladi-velikov — DISMISSED** [review]
Awesome work, Mirela! Our first Java FaaS function!

You have addressed all my comments shared offline. Let's have one more review before merging this pull request.

File: examples/java/veeam-vm-backup/README.MD
Comment (embano1) [reply examples/java/veeam-vm-backup/README.MD]: Can you please change the Java version for the image then to v11 as well please? v8 is marked deprecated and it would be great to have a function using an up to date image.

Comment (embano1) [reply examples/java/veeam-vm-backup/README.MD]: After the change, please also update the README accordingly.

File: examples/java/veeam-vm-backup/README.MD
Comment (embano1) [inline examples/java/veeam-vm-backup/README.MD:76]: This section is only relevant for developers/if someone wants to build the image vs just deploying it. Can you please break these sections up (e.g. deploy/develop) so it is not confusing for the reader which likely just wants to deploy the (pre-built) image.

File: examples/java/veeam-vm-backup/README.MD
Comment (embano1) [inline examples/java/veeam-vm-backup/README.MD:97]: Please see comments above.

File: examples/java/veeam-vm-backup/README.MD
Comment (embano1) [inline examples/java/veeam-vm-backup/README.MD:100]:
```suggestion
faas-cli template pull # only required during the first build/deployment
```

**Review by embano1 — CHANGES_REQUESTED** [review]
Thanks for the update. Just some minor tasks left. Once done and before the merge can you please add a line-break between the commit title and message? Otherwise it's messing up the commit message.

**Review by vladi-velikov — APPROVED** [review]

File: examples/java/veeam-vm-backup/README.MD
Comment (mirela-manoleva) [reply examples/java/veeam-vm-backup/README.MD]: Done.

File: examples/java/veeam-vm-backup/README.MD
Comment (mirela-manoleva) [reply examples/java/veeam-vm-backup/README.MD]: Done!

File: examples/java/veeam-vm-backup/README.MD
Comment (mirela-manoleva) [reply examples/java/veeam-vm-backup/README.MD:79]: Done!

**Comment by mirela-manoleva** [comment]
> Thanks for the update. Just some minor tasks left. Once done and before the merge can you please add a line-break between the commit title and message? Otherwise it's messing up the commit message.

All requested changes have been done. If the function is ready for merging, could you please publish the docker image: https://hub.docker.com/repository/docker/manoleva/veba-java-veeam-vm-backup

File: examples/java/veeam-vm-backup/README.MD
Comment (embano1) [inline examples/java/veeam-vm-backup/README.MD:7]:
```suggestion
The function is written in Java - Java 11 by default, but Java 8 is supported, too. It gets called on a VM state changed event. Then it [logs on](https://helpcenter.veeam.com/docs/backup/rest/post_sessionmngr.html?ver=100) to the Veeam's server, finds the hierarchyRoot ([per Veeam documentation](https://helpcenter.veeam.com/docs/backup/rest/get_hierarchyroots.html?ver=100)) and constructs the VM reference per Veeam format `urn:VMware:vm:VEEAM_HOST_ID:VM_VALUE`. For example: `urn:VMware:vm:3f3fe831-79b8-4123-9ff9-3d5df7e89c9f.vm-28`), which is different from the VMware one, so that the actual [backup](https://helpcenter.veeam.com/docs/backup/rest/post_backupservers_id_quickbackup.html?ver=100) succeeds.
```

File: examples/java/veeam-vm-backup/README.MD
Comment (embano1) [inline examples/java/veeam-vm-backup/README.MD:57]:
```suggestion
> **Note:** Delete the local `veeamconfig.toml` after you're done with this exercise to not expose this sensitive information.
```

File: examples/java/veeam-vm-backup/README.MD
Comment (embano1) [inline examples/java/veeam-vm-backup/README.MD:93]:
```suggestion
> **Note:** If you don't see a backup for the VM verify that you correctly followed each step above, IPs/FQDNs and credentials are correct and see the [troubleshooting](#troubleshooting) section below.
```

**Review by embano1 — APPROVED** [review]
LGTM, just some minor markdown changes proposed. Feel free to accept and squash merge into your commit. Then we're good to merge.

**Review by embano1 — APPROVED** [review]

**Pull request merged**
